lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -706,28 +717,79 @@ class TypeRefinementContextBuilder : private ASTWalker {
     return D->getSourceRange();
   }
 
+  // Creates an implicit decl TRC specifying the deployment
+  // target for `range` in decl `D`.
+  TypeRefinementContext *
+  createImplicitDeclContextForDeploymentTarget(Decl *D, SourceRange range){
+    AvailabilityContext Availability =
+        AvailabilityContext::forDeploymentTarget(Context);
+    Availability.intersectWith(getCurrentTRC()->getAvailabilityInfo());
+
+    return TypeRefinementContext::createForDeclImplicit(
+        Context, D, getCurrentTRC(), Availability, range);
+  }
+
+  /// Build contexts for a VarDecl with the given initializer.
+  void buildContextsForPatternBindingDecl(PatternBindingDecl *pattern) {
+    // Build contexts for each of the pattern entries.
+    for (unsigned index : range(pattern->getNumPatternEntries())) {
+      auto var = pattern->getAnchoringVarDecl(index);
+      if (!var)
+        continue;
+
+      // Var decls may have associated pattern binding decls or property wrappers
+      // with init expressions. Those expressions need to be constrained to the
+      // deployment target unless they are exposed to clients.
+      if (!var->hasInitialValue() || var->isInitExposedToClients())
+        continue;
+
+      auto *initExpr = pattern->getInit(index);
+      if (initExpr && !initExpr->isImplicit()) {
+        assert(initExpr->getSourceRange().isValid());
+
+        // Create a TRC for the init written in the source. The ASTWalker
+        // won't visit these expressions so instead of pushing these onto the
+        // stack we build them directly.
+        auto *initTRC = createImplicitDeclContextForDeploymentTarget(
+            var, initExpr->getSourceRange());
+        TypeRefinementContextBuilder(initTRC, Context).build(initExpr);
+      }
+    }
+
+    // Ideally any init expression would be returned by `getInit()` above.
+    // However, for property wrappers it doesn't get populated until
+    // typechecking completes (which is too late). Instead, we find the
+    // the property wrapper attribute and use its source range to create a
+    // TRC for the initializer expression.
+    //
+    // FIXME: Since we don't have an expression here, we can't build out its
```

Comment:
Is this comment no longer true, since we can now lazily expand the TRC for the init expr? I'm having a little trouble following whether this is handled now, but it sounds like it might be from a comment in the request implementation.

Reply:
I don't know that this has fully addressed the issue, because it seems like we'd need to add the TRCs for the property wrapper attributes (below). I'd need to dig into rdar://77841331 to see what it would take to fix that.

Reply:
Aside from the bugs I've added, I suspect this is related... this fails to type-check both before and after my PR:

```swift
@available(macOS 10.15, *)
func f() -> Int { 17 }

class X {
  @available(macOS 10.14, *)
  private static let value: Int = {
    if #available(macOS 10.15, *) {
      return f() // produces an error here saying that f() is not available
    }

    return 0
  }()
}
```

Reply:
Ok, thanks for checking. I'll take a look at fixing that eventually!

Reply:
I changed the approach a little bit to in [this commit](https://github.com/apple/swift/pull/67642/commits/51b7b4d9658f793464f7041d912f54de3498c50a) deal with a failure in the source compatibility suite, which also fixes the example I gave above---an existing bug that's now fixed by my PR. It doesn't address rdar://77841331, though